data_model/src/trigger.rs

```diff
@@ -244,6 +249,57 @@ pub mod action {
         }
     }
 
+    mod candidate {
+        use parity_scale_codec::Input;
+
+        use super::*;
+
+        #[derive(Decode, Deserialize)]
+        pub(super) struct ActionCandidate {
+            pub executable: Executable,
+            pub repeats: Repeats,
+            pub authority: AccountId,
+            pub filter: EventFilterBox,
+            pub metadata: Metadata,
+        }
+
+        impl ActionCandidate {
+            pub(super) fn validate(self) -> Result<Action, &'static str> {
+                if matches!(self.filter, EventFilterBox::TriggerCompleted(_)) {
+                    return Err("TriggerCompleted cannot be used as filter for triggering actions");
+                }
```

Comment:
Just as an idea: Iroha can have an internal newtype around `EventFilterBox` which ensures it never holds the `TriggerCompleted` variant.